### ISSUE TITLE
cannot reference the internal Promise made by eventual-send?

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#886]
While trying to write the tests in #859, I came across an unexpected behavior in the way SwingSet uses eventual-send. I think it's behaving as designed, but it feels unfortunate, and I'm trying to see if I can work around it or if this is likely to cause problems in the future.

When code inside a vat sends a message to somebody, it uses something like this:

```js
const p1 = E(bob).foo(); // equivalent to p1 = bob~.foo()
```

`bob`, in this case, is a "Presence": an empty object which is internally associated with a particular `HandledPromise`, which has a particular handler. This handler knows about the `o-NN` reference that represents "bob" on some other vat, and it knows how to make a `syscall.send()` with the method name and serialized arguments. This handler is defined in liveslots (in `makeQueued(slot)`, called by `convertSlotToVal()` when dealing with object references), and is built something like this:

```js
const handler = {
  applyMethod(_0, prop, args) {
    const p0 = queueMessage(slot, prop, args); // do syscall.send
    return p0;
  },
};
let resolveWithPresence
const hp = new HandledPromise((_, _, resPres => { resolveWithPresence = resPres }),
                              handler);
const presence = resolveWithPresence(handler); // fulfilledHandler
// discard hp, return presence
```

`queueMessage` creates a new promise identifier (a `p+NN` string) and uses it as the `result=` argument of the `syscall.send`. This tells the kernel how to notify the sending vat about the result of the message send: at some point in the future, this vat will receive a `dispatch.notifyFullToPresence` of some sort which references the same `p+NN` string, and the liveslots layer will resolve `p0`.

We add `p0` into the `valToSlot`/`slotToVal` tables, so in the off chance that other vats talk back to this one about this result promise, those references will be deserialized back into the correct promise.

What I had previously believed was that `p1 === p0`. I'm now realizing that this couldn't be true, because the `HandledPromise` handler is not trusted with synchronous behavior, and `p1` has been long-since created and returned by the time the handler is invoked to create `p0`. My current (tenuous) belief is that `p1` is eventually resolved to `p0`, which gives `p1` the same handler as `p0`, allowing pipelined sends to work (`p1~.bar()`). But they are distinct Promises, and most importantly `p1` is not added to `valToSlot` or `slotToVal`.

The consequence for the tests I'm writing is that one of the cases simply cannot be reached by liveslots-based vats. I wanted to exercise a case where a vat sends a message, creating result promise p1, and then sends `p1` as an argument in a subsequent message. I was expecting that the kernel will see the same promise identifier for both the result and the subsequent argument, and I was looking to test how the kernel (and other liveslots instances) reacts to that promise being resolved. But what I'm seeing is the result promise is serialized as `p+5`, while the later argument gets `p+6`. Now I'm realizing that the handler's internal `p0` is being registered as `p+5` during `queueMessage`, but the application-level code receives a distinct `p1`, and when that meets the serializer, it is assigned a brand new number.

I don't think this is affecting the ability to pipeline messages (`p1~.bar()`) because the handler is forwarded. However I think it *is* introducing an additional turn of delay into the transmission, which could be an ordering issue.

But it means that liveslots-based vats cannot reference their own result promises without external help. I think this will introduce addtional cranks (and roundtrips between vats) when promises are passed as arguments. E.g.:

```js
Alice: p1 = bob~.foo(); carol~.bar(p1)
Bob: function foo() { return 'f'; }
Carol: function bar(p1) { p1.then(f => console.log('carol got f')) }
```

In that example, I'd really prefer carol to receive the same promise ID that Bob is going to resolve. If that happened, then when Bob finally resolves it, the kernel will notify everyone subscribed to hear about it (both Alice and Carol), and Carol is notified pretty directly.

But if Alice's `p1` is not the same as the `result=` that Bob gets, then we have two distinct Promises in play. The first is subscribed-to by Alice, and decided-by/resolved-by Bob. The second is subscribed-to by Carol, and decided-by/resolved-by Alice. The resolution has to go from Bob to Alice first, and then Alice reacts to it by resolving the promise that Carol is watching.

I need to run some more tests to see if this new understanding is accurate, but it would explain what I've seen so far. @FUDCo I think it would also explain the additional turns we were seeing in that test.

@erights @michaelfig does this sound accurate? Does it also seem like an unfortunate performance hit?

**Comment by erights** [comment]
ok. I think the following is kludgy. It would need to be added to the eventual-send shim and therefore the proposal. It makes the proposal less elegant. But let's see if it actually solves the problem.
Say that the handler traps, at least `get` and `applyMethod`, took an additional argument they were free to ignore. This is the promise that the corresponding static method (which has no such extra parameter) had already returned.
Whatever the handler trap returns, this will be used to resolve that promise. Thus, if the trap returns a promise, the already returned promise will be forwarded to that one, after which they are behaviorally equivalent.
In that case, whatever registration the handler might do on the promise it returns, it can equally well do with the promise it gets as that extra parameter.
The two promises are behaviorally equivalent. But the extra parameter promise is the one seen by user code. Thus, it is the one whose identity might be seen again.

**Comment by warner** [comment]
Would it be legal/sensible for the handler to return the previously-generated Promise that it was given? (to enable that, the implementation would probably have to take care to not attempt to resolve a Promise to itself)

**Comment by erights** [comment]
But the handler is where the value to resolve it to comes from. A correct handler should always return the thing to resolve the promise to.

**Comment by erights** [comment]
Fixed by #897
